### ISSUE TITLE
NODEFS.flagsForNode does not support O_NOCTTY

### ISSUE THREAD

**Issue by zhuowei** [emscripten-core/emscripten#11927]
`gzip` [uses O_NOCTTY](https://fossies.org/linux/gzip/gzip.c#l_1294) when opening files; Emscripten [doesn't support](https://github.com/emscripten-core/emscripten/blob/49029d44cbef0f535fcd48445300808c7efbdb10/src/library_nodefs.js#L19) that flag.

To reproduce:

Download http://archive.ubuntu.com/ubuntu/pool/main/g/gzip/gzip_1.10.orig.tar.gz
Build and run with:

```
mkdir build
cd build
emconfigure ../configure --disable-gcc-warnings TIME_T_32_BIT_OK=yes LDFLAGS="-O2 -s NODERAWFS=1"
emmake make -j4
node gzip /any/file.txt
```

Expected: the Emscripten'd gzip compresses the file
Actual:

```
gzip: /any/file.txt: Invalid argument
```

Versions:

```
   node-12.18.1-64bit
   python-3.7.4-2-64bit
   releases-upstream-5974288502aab433d45f53511e961aaca4079d86-64bit
```

**Comment by algestam** [comment]
A PR (#12492) that adds the `O_NOCTTY` flag has been submitted. The change partly resolves the issue with gzip as files are getting compressed but another error is emitted once the file has been compressed (`gzip: testfile.gz: Not a directory`). The reason to why this happens seems to be that a call to `futimens` fails. I'm not sure if this is related to the `O_NOCTTY` flag or if it's related to something else.

**Comment by algestam** [comment]
Have investigated the issue with gzip outputting `gzip: testfile.gz: Not a directory` a bit further.

TL;DR - The issue is not related to the `O_NOCTTY` flag missing (which has been resolved earlier) but is most likely related to an issue with setting file timestamps.

Build and run:

    mkdir build
    cd build
    emconfigure ../configure --disable-gcc-warnings TIME_T_32_BIT_OK=yes LDFLAGS="-O2 -s NODERAWFS=1"
    emmake make -j4
    echo "test" > testfile
    node gzip testfile

Output:

    gzip: testfile.gz: Not a directory

However, a compressed file is generated:

    $ ls testfile*
    testfile  testfile.gz

Debugging with `node --inspect gzip testfile` indicates that an exception is thrown while restoring the file timestamps (using futimens/utimensat). The exception message is `"ENOTDIR: not a directory, utime '/path/to/build/testfile.gz/'"` and this is the stacktrace:

    handleErrorFromBinding (fs.js:114)
    utimesSync (fs.js:1113)
    utime (gzip:2576)
    (anonymous) (gzip:5007)
    ___sys_utimensat (gzip:4684)
    Module._main (gzip:5093)
    callMain (gzip:5194)
    doRun (gzip:5253)
    run (gzip:5268)
    runCaller (gzip:5172)
    removeRunDependency (gzip:1293)
    receiveInstance (gzip:1430)
    receiveInstantiatedSource (gzip:1441)

It seems strange though that the path argument used to update the file timestamp is ending with a slash (`/path/to/build/testfile.gz/`). The path is calculated from the file descriptor in the `__sys_utimensat()` function in `library_syscall.js` using `SYSCALLS.calculateAt(dirfd, path)`.

Simply stripping the the last slash (if any) from the `path` variable  in `__sys_utimensat()` after the path has been calculated resolves this specific issue but I'm not sure if that's the best solution here or if the issue is in the `calculateAt()` function.

I'm happy to look further into this but need advice from someone with a better understanding of this :)

**Comment by sbc100** [comment]
Looks like a real bug... sounds like you are well placed to fix it ;-)

**Comment by sbc100** [comment]
Also it looks like this particular has been fixed now?   Can we close it?

**Comment by algestam** [comment]
Yeah, I think this issue can be closed as the `O_NOCTTY` flag now is supported. Will continue looking into the other problem found.
